narwhals/_pandas_like/utils.py

```diff
@@ -412,7 +413,15 @@ def translate_dtype(dtype: Any) -> DType:
     if str(dtype) == "date32[day][pyarrow]":
         return dtypes.Date()
     if dtype == "object":
-        return dtypes.String()
+        if (idx := column.first_valid_index()) is not None and isinstance(
+            column.loc[idx], str
+        ):
+            # Infer based on first non-missing value.
+            # For pandas pre 3.0, this isn't perfect.
+            # After pandas 3.0, pandas has a dedicated string dtype
+            # which is inferred by default.
+            return dtypes.String()
+        return dtypes.Object()
```

Comment:
Not sure if the comment is related, but shouldn't all non-null/nan elements be str?

Reply:
you can store anything in object dtype, it doesn't have to be strings

Reply:
That's exactly my point, _all_ elements should be str, not just the first one.

Reproducible example with the above implementation:

```py
class Bar:
    ...

df = pd.DataFrame({"a": ["foo", Bar()]}).astype(object)
result = nw.from_native(df).schema
result["a"] == nw.String
```

```terminal
True
```

Reply:
ah sure, I see what you mean

true. this is just an approximation, we're just matching what Polars does here https://docs.pola.rs/api/python/stable/reference/series/index.html